### PR TITLE
feat: provide signing config separately from the main spec

### LINKED ISSUE #282: [REQ] add option to skip signing
### What kind of request is this?

Improvement of existing experience

### What is your request or suggestion?

When we specify the signing block for a target in a `dalec` spec, it defaults to calling the signer. This can be inconvenient when attempting to test builds in a local environment, especially when we are not aiming to test signing.

It would be beneficial to have a `build-arg` or an equivalent flag/option to bypass the signing call.

Maybe, we might want to consider moving the signing block to a `build-arg` or flag. From my perspective, having the dalec spec and the signing block together makes it harder to share the specs directly. We would need to inform the recipient that they need to adjust the signing block to use their own signer image or to remove it. Alternatively, we would need to modify the spec and delete the signing block before sharing it.

### Are you willing to submit PRs to contribute to this feature request?

- [ ] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by pmengelbert**
**What this PR does / why we need it**:
This PR allows you to specify the signing config separately from the main spec.

You can do this in one of three ways:
1. Provide a build context with the exact name `dalec_signing_config` which points to a directory containing a file with the exact name `dalec_signing_config.yml`
1. Provide a build context as above, but with a custom path for the config, by using the build arg `DALEC_SIGNING_CONFIG_PATH`. This path will be relative to the root of the supplied build context.
1. Provide a custom path for the config on your local filesystem, by using the build arg `DALEC_SIGNING_CONFIG_PATH`. This path can be absolute, or relative to the context location from the `docker build` invocation.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #282 

**Special notes for your reviewer**:

### CODE REVIEW

File: frontend/request.go
Comment (cpuguy83) [inline frontend/request.go:149]: nit: no need for `try` here.

File: frontend/request.go
Comment (cpuguy83) [reply frontend/request.go:194]: To simplify this down further, maybe there is _no_ implicit context name we will look for.
Instead the build-arg for the filename must always be provided, and if the client wants to use a custom context they need to tell us that in another build arg (e.g. `DALEC_SIGNING_CONTEXT_NAME`)

File: frontend/request.go
Comment (cpuguy83) [inline frontend/request.go:272]: Can you make the base-key a const?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go:152]: > This is fairly expensive to create.

Surprising -- why is that?

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go:167]: done

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go]: done

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go]: done

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go]: Yes

Comment (pmengelbert) [reply frontend/request.go]: done

File: frontend/request.go
Comment (cpuguy83) [inline frontend/request.go:201]: Comment looks like bad copy/psate since this is checking the context.

File: test/fixtures/signer/sign_config.yml
Comment (cpuguy83) [inline test/fixtures/signer/sign_config.yml:1]: Doesn't look like we need this file?

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:22]: This is a fairly generic function name that's only minimally useful.
`newSimpleSpec`?

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:52]: This mostly looks like a dup of the actual `newSpec`?

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:210]: Can we trim these paths down since they aren't needed?

File: frontend/request.go
Comment (cpuguy83) [inline frontend/request.go:192]: After #322 is merged, we should add a warning when the build-args are overwriting the signing in the spec.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: frontend/request.go
Comment (cpuguy83) [reply frontend/request.go]: > Surprising -- why is that?

It is eagerly initializing a bunch of stuff on `NewClient`

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go:192]: sgtm, will wait till that one's merged

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go]: done

File: test/fixtures/signer/sign_config.yml
Comment (pmengelbert) [reply test/fixtures/signer/sign_config.yml]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: It set up a signing config. I refactored it to deduplicate.

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: done

File: frontend/request.go
Comment (cpuguy83) [reply frontend/request.go:194]: That one is in now. Let's also make sure to bring back the tests that check to make sure the warning is present 😅 🚀

Comment (pmengelbert) [reply frontend/request.go:189]: Merged that, updated this

Comment (pmengelbert) [reply frontend/request.go:189]: done

Comment (cpuguy83) [reply frontend/request.go:189]: Looks like tests are missing the checks for the warning messages (see: `WithSolveStatusFn`)

Comment (pmengelbert) [reply frontend/request.go:189]: I need to slow down!

Comment (pmengelbert) [reply frontend/request.go:189]: done

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:150]: The test is never checking this value.

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:88]: This could use a similar check to see if the warning is provided.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go:150]: thanks

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go:88]: done -- had to implement the warning to get the test to pass :)

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go:165]: done

File: frontend/request.go
Comment (cpuguy83) [inline frontend/request.go:191]: Can we restructure this like to reduce indentation?

```go
cfgPath := getUserSignConfigPath(client)
if cfgPath != "" {
  // handle
  return updatedSt, nil
}

cfg, warning := spec.GetSigner(targetKey)
// handle the rest
```

File: helpers.go
Comment (cpuguy83) [inline helpers.go:378]: Instead this could return a bool to help simplify understanding what's going on here.
Also would need a doc string to explain what the 2nd return is for.
Since `Warning` isn't really an error this feels a bit strange.

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:244]: Also needs a warning check.

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:294]: Needs warning check.

File: test/signing_test.go
Comment (cpuguy83) [inline test/signing_test.go:317]: Needs warning check.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: helpers.go
Comment (pmengelbert) [reply helpers.go:378]: We changed this to remove the bool which is why I went with this. But I'll put the bool back

Comment (cpuguy83) [reply helpers.go:378]: The old case was returning `true` if `PackageSIgner` was not nil.
I think what we want is to return a bool when the package signer in the target is superseding the base?

File: frontend/request.go
Comment (pmengelbert) [reply frontend/request.go]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go]: done

File: test/signing_test.go
Comment (pmengelbert) [reply test/signing_test.go:256]: done

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**
